### ISSUE TITLE
proto: support UTF-8 in TextMarshaler

### ISSUE THREAD

**Issue by lyfunny** [golang/protobuf#572]
```
fmt.Println(rsp.String())
```
i got 
```
list: <
  uuid: "171001669"
  roomid: "78901"
  nick: "u3d002\350\211\272\344\272\272"
  label: "\350\213\261\351\233\204\350\201\224\347\233\237"
>
```
but i want to see bellow in log file，so can you offer a Utf8String() like Utf8DebugString() in c++ protobuf?
```
list: <
  uuid: "171001669"
  roomid: "78901"
  nick: "如何显示中文逆天"
  label: "看不懂这些乱码！"
```

**Comment by dsnet** [comment]
The "text serialization" of protos was originally defined to be ASCII only. For that reason, it always escapes non-ASCII. This is the behavior of C++ and we copied it in Go to remain compatible. However, I noticed that the [C++ implementation provides an option to do this](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format#TextFormat.Printer.SetUseUtf8StringEscaping.details). We could add an option to [`TextMarshaler`](https://godoc.org/github.com/golang/protobuf/proto#TextMarshaler) to support this.

**Comment by ashi009** [comment]
Is there a timetable for releasing v2 API? It sounds like a major release that won't come anytime soon.

This sounds like a minor change, shall we implement it now? 

Also, should it be the default text marshaling behavior in the future?

**Comment by cybrcodr** [comment]
v2 will take a while, we're hoping we'll have something by end of Q1 2019.  We'll consider 
 UTF8 as default for v2 API, or even just not have option to output encoded ASCII.

Implementing it right now means adding an option that we'll have to continue supporting in v1, not quite keen on that.

**Comment by ashi009** [comment]
Sounds fair.

> Implementing it right now means adding an option that we'll have to continue supporting in v1, not quite keen on that.

IMHO, the transition from v1 to v2 API won’t be easy and will take years. So the maintenance burden exists already. From this perspective adding this to the current API has its value.

**Comment by dsnet** [comment]
The v1 API will be implemented in terms of the v2 API. I fully expect migration to v2 to take years, but the benefits of v2 will be visible in v1 much sooner. That said, we don't want to continue investing effort into v1 that will make v2 harder.

**Comment by dsnet** [comment]
This is fixed in the [v1.20.0 release](https://github.com/protocolbuffers/protobuf-go/releases/tag/v1.20.0) where the `prototext` package now emits strings as UTF-8 by default. There is currently no option to emit pure-ASCII. If a need for this arises, we can add an option for that.

**Comment by xialu4820723** [comment]
There is an other way to print UTF-8 normally.

```golang
import "github.com/golang/protobuf/jsonpb"

str, _ := new(jsonpb.Marshaler).MarshalToString(proMessage)
fmt.Println(str)
```
